src/servers/src/auth.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2022 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod mysql;
+pub mod postgres;
+pub mod pwd;
+
+use std::sync::Arc;
+
+use serde::{Deserialize, Serialize};
+
+use self::mysql::MysqlAuthPlugin;
+use self::postgres::PgAuthPlugin;
+use crate::error::Result;
+
+pub const DEFAULT_USERNAME: &str = "greptime";
+
+#[async_trait::async_trait]
+pub trait UserProvider: Send + Sync {
+    fn name(&self) -> String;
+
+    async fn get_user_info(&self, identity: &Identity) -> Result<Option<UserInfo>>;
+}
+
+pub type UserProviderRef = Arc<dyn UserProvider>;
+
+type Username = String;
+type HostOrIp = String;
+
+#[derive(Debug)]
+pub enum Identity {
+    UserId(Username, Option<HostOrIp>),
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct UserInfo {
+    username: String,
+    auth_methods: Vec<AuthMethod>,
+    // TODO(fys): maybe contain some user permission information here
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum AuthMethod {
+    PlainText(Vec<u8>),
+    DoubleSha1(Vec<u8>),
+}
+
+impl UserInfo {
+    pub fn new(username: impl Into<String>, auth_methods: Vec<AuthMethod>) -> Self {
+        Self {
+            username: username.into(),
+            auth_methods,
+        }
+    }
+
+    pub fn get_username(&self) -> String {
+        self.username.clone()
+    }
```

Comment:
Prefer not to use "`get`" prefix in getter's name in Rust; also better to return ref instead of cloning a new one:

```suggestion
    pub fn username(&self) -> &String {
        &self.username
    }
```